pkg/sql/schemachanger/schemachanger_test: TestConcurrentSchemaChanges failed

pkg/sql/schemachanger/schemachanger_test.TestConcurrentSchemaChanges [failed](https://tanzanite.cluster.engflow.com/invocations/default/a8072aa6-f595-4364-abc8-ea3705495149?testReportRun=10&testReportShard=24&testReportAttempt=1#targets-Ly9wa2cvc3FsL3NjaGVtYWNoYW5nZXI6c2NoZW1hY2hhbmdlcl90ZXN0) on release-24.1 @ [957fad03d6876049dc0faef2217858e5ea92f5ca](https://github.com/cockroachdb/cockroach/commits/957fad03d6876049dc0faef2217858e5ea92f5ca):


```
    schemachanger_test.go:760: RENAME TABLE TO t_88
    schemachanger_test.go:760: RENAME TABLE TO t_89
    schemachanger_test.go:787: Parent database or schema or table is renamed or column already exists; skipping this column addition.
    schemachanger_test.go:787: Parent database or schema or table is renamed or column already exists; skipping this column addition.
    schemachanger_test.go:732: RENAME SCHEMA TO testsc_90
    schemachanger_test.go:783: ADD COLUMN col_93 TO testdb_76.testsc_90.t_89
    schemachanger_test.go:783: ADD COLUMN col_96 TO testdb_76.testsc_90.t_89
    schemachanger_test.go:807: DROP COLUMN col_93 FROM testdb_76.testsc_90.t_89
    schemachanger_test.go:838: Parent database or schema or table is renamed or column is dropped or index already exists; skipping this index creation.
    schemachanger_test.go:760: RENAME TABLE TO t_98
    schemachanger_test.go:760: RENAME TABLE TO t_100
    schemachanger_test.go:783: ADD COLUMN col_101 TO testdb_76.testsc_90.t_100
    schemachanger_test.go:811: Parent database or schema or table is renamed; skipping this column removal.
    schemachanger_test.go:698: DROP DATABASE testdb_76
    schemachanger_test.go:739: Parent database is renamed; skipping this schema renaming.
    schemachanger_test.go:698: DROP DATABASE testdb_76
    schemachanger_test.go:698: DROP DATABASE testdb_76
    schemachanger_test.go:783: ADD COLUMN col_106 TO testdb_76.testsc_90.t_100
    schemachanger_test.go:807: DROP COLUMN col_106 FROM testdb_76.testsc_90.t_100
    schemachanger_test.go:783: ADD COLUMN col_110 TO testdb_76.testsc_90.t_100
    schemachanger_test.go:762: DROP TABLE t_109
    schemachanger_test.go:783: ADD COLUMN col_112 TO testdb_76.testsc_90.t_100
    schemachanger_test.go:807: DROP COLUMN col_112 FROM testdb_76.testsc_90.t_100
    schemachanger_test.go:783: ADD COLUMN col_116 TO testdb_76.testsc_90.t_100
    schemachanger_test.go:838: Parent database or schema or table is renamed or column is dropped or index already exists; skipping this index creation.
    schemachanger_test.go:807: DROP COLUMN col_116 FROM testdb_76.testsc_90.t_100
    schemachanger_test.go:734: DROP SCHEMA TO testsc_90
    schemachanger_test.go:783: ADD COLUMN col_118 TO testdb_76.testsc_90.t_100
    schemachanger_test.go:807: DROP COLUMN col_118 FROM testdb_76.testsc_90.t_100
    schemachanger_test.go:698: DROP DATABASE testdb_76
    schemachanger_test.go:838: Parent database or schema or table is renamed or column is dropped or index already exists; skipping this index creation.
    schemachanger_test.go:783: ADD COLUMN col_122 TO testdb_76.testsc_90.t_100
    schemachanger_test.go:698: DROP DATABASE testdb_76
    schemachanger_test.go:709: RENAME DATABASE TO testdb_125
    schemachanger_test.go:709: RENAME DATABASE TO testdb_126
    schemachanger_test.go:679: rename-tbl-worker encounters error driver: bad connection; signal to main routine and finish working
    schemachanger_test.go:675: rename-schema-worker is signaled to finish work
    schemachanger_test.go:675: rename-db-worker is signaled to finish work
    schemachanger_test.go:675: add-column-worker is signaled to finish work
    schemachanger_test.go:675: drop-index-worker is signaled to finish work
    schemachanger_test.go:675: create-index-worker is signaled to finish work
    schemachanger_test.go:675: drop-column-worker is signaled to finish work
    schemachanger_test.go:869: 
        	Error Trace:	pkg/sql/schemachanger/schemachanger_test_test/pkg/sql/schemachanger/schemachanger_test.go:869
        	Error:      	Received unexpected error:
        	            	driver: bad connection
        	Test:       	TestConcurrentSchemaChanges
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestConcurrentSchemaChanges2973834955
--- FAIL: TestConcurrentSchemaChanges (40.92s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=10</code>
 - <code>shard=24</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConcurrentSchemaChanges.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-41392

pkg/sql/schemachanger/schemachanger_test.TestConcurrentSchemaChanges [failed](https://tanzanite.cluster.engflow.com/invocations/default/d39d034b-eb7c-4710-a34d-5ddc2a8602ab?testReportRun=25&testReportShard=24&testReportAttempt=1#targets-Ly9wa2cvc3FsL3NjaGVtYWNoYW5nZXI6c2NoZW1hY2hhbmdlcl90ZXN0) on release-24.1 @ [c977778dc8419323431a990baf3d702db4fd584b](https://github.com/cockroachdb/cockroach/commits/c977778dc8419323431a990baf3d702db4fd584b):


```
    schemachanger_test.go:783: ADD COLUMN col_490 TO testdb_469.testsc_482.t_489
    schemachanger_test.go:830: CREATE INDEX idx_493 ON testdb_469.testsc_491.t_489(col_490)
    schemachanger_test.go:830: CREATE INDEX idx_495 ON testdb_469.testsc_491.t_489(col_490)
    schemachanger_test.go:856: DROP INDEX idx_495 FROM testdb_469.testsc_491.t_489
    schemachanger_test.go:830: CREATE INDEX idx_496 ON testdb_469.testsc_491.t_489(col_490)
    schemachanger_test.go:856: DROP INDEX idx_493 FROM testdb_469.testsc_491.t_489
    schemachanger_test.go:807: DROP COLUMN col_490 FROM testdb_469.testsc_491.t_489
    schemachanger_test.go:783: ADD COLUMN col_494 TO testdb_469.testsc_491.t_489
    schemachanger_test.go:734: DROP SCHEMA TO testsc_491
    schemachanger_test.go:732: RENAME SCHEMA TO testsc_500
    schemachanger_test.go:732: RENAME SCHEMA TO testsc_501
    schemachanger_test.go:767: Parent database or schema is renamed; skipping this table renaming.
    schemachanger_test.go:732: RENAME SCHEMA TO testsc_502
    schemachanger_test.go:734: DROP SCHEMA TO testsc_502
    schemachanger_test.go:783: ADD COLUMN col_505 TO testdb_469.testsc_502.t_489
    schemachanger_test.go:783: ADD COLUMN col_506 TO testdb_469.testsc_502.t_489
    schemachanger_test.go:830: CREATE INDEX idx_507 ON testdb_469.testsc_502.t_489(col_506)
    schemachanger_test.go:807: DROP COLUMN col_505 FROM testdb_469.testsc_502.t_489
    schemachanger_test.go:830: CREATE INDEX idx_509 ON testdb_469.testsc_502.t_489(col_506)
    schemachanger_test.go:856: DROP INDEX idx_507 FROM testdb_469.testsc_502.t_489
    schemachanger_test.go:783: ADD COLUMN col_508 TO testdb_469.testsc_502.t_489
    schemachanger_test.go:856: DROP INDEX idx_509 FROM testdb_469.testsc_502.t_489
    schemachanger_test.go:807: DROP COLUMN col_506 FROM testdb_469.testsc_502.t_489
    schemachanger_test.go:732: RENAME SCHEMA TO testsc_513
    schemachanger_test.go:838: Parent database or schema or table is renamed or column is dropped or index already exists; skipping this index creation.
    schemachanger_test.go:830: CREATE INDEX idx_514 ON testdb_469.testsc_513.t_489(col_508)
    schemachanger_test.go:856: DROP INDEX idx_514 FROM testdb_469.testsc_513.t_489
    schemachanger_test.go:787: Parent database or schema or table is renamed or column already exists; skipping this column addition.
    schemachanger_test.go:783: ADD COLUMN col_515 TO testdb_469.testsc_513.t_489
    schemachanger_test.go:807: DROP COLUMN col_508 FROM testdb_469.testsc_513.t_489
    schemachanger_test.go:830: CREATE INDEX idx_517 ON testdb_469.testsc_513.t_489(col_515)
    schemachanger_test.go:767: Parent database or schema is renamed; skipping this table renaming.
    schemachanger_test.go:762: DROP TABLE t_518
    schemachanger_test.go:760: RENAME TABLE TO t_519
    schemachanger_test.go:679: rename-db-worker encounters error driver: bad connection; signal to main routine and finish working
    schemachanger_test.go:675: rename-schema-worker is signaled to finish work
    schemachanger_test.go:675: drop-column-worker is signaled to finish work
    schemachanger_test.go:675: rename-tbl-worker is signaled to finish work
    schemachanger_test.go:675: drop-index-worker is signaled to finish work
    schemachanger_test.go:675: create-index-worker is signaled to finish work
    schemachanger_test.go:787: Parent database or schema or table is renamed or column already exists; skipping this column addition.
    schemachanger_test.go:675: add-column-worker is signaled to finish work
    schemachanger_test.go:869: 
        	Error Trace:	pkg/sql/schemachanger/schemachanger_test_test/pkg/sql/schemachanger/schemachanger_test.go:869
        	Error:      	Received unexpected error:
        	            	driver: bad connection
        	Test:       	TestConcurrentSchemaChanges
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestConcurrentSchemaChanges1412864246
--- FAIL: TestConcurrentSchemaChanges (162.88s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=25</code>
 - <code>shard=24</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConcurrentSchemaChanges.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

pkg/sql/schemachanger/schemachanger_test.TestConcurrentSchemaChanges [failed](https://tanzanite.cluster.engflow.com/invocations/default/c66f1a0e-f7f6-40f3-b093-bc42e60d982e?testReportRun=8&testReportShard=24&testReportAttempt=1#targets-Ly9wa2cvc3FsL3NjaGVtYWNoYW5nZXI6c2NoZW1hY2hhbmdlcl90ZXN0) on release-24.1 @ [67a92660d15a856d40f644f2fceb0a47923ea323](https://github.com/cockroachdb/cockroach/commits/67a92660d15a856d40f644f2fceb0a47923ea323):


```
    schemachanger_test.go:811: Parent database or schema or table is renamed; skipping this column removal.
    schemachanger_test.go:698: DROP DATABASE testdb_316
    schemachanger_test.go:783: ADD COLUMN col_369 TO testdb_316.testsc_349.t_345
    schemachanger_test.go:783: ADD COLUMN col_370 TO testdb_316.testsc_349.t_345
    schemachanger_test.go:807: DROP COLUMN col_369 FROM testdb_316.testsc_349.t_345
    schemachanger_test.go:830: CREATE INDEX idx_371 ON testdb_316.testsc_349.t_345(col_370)
    schemachanger_test.go:783: ADD COLUMN col_372 TO testdb_316.testsc_349.t_345
    schemachanger_test.go:762: DROP TABLE t_373
    schemachanger_test.go:762: DROP TABLE t_374
    schemachanger_test.go:811: Parent database or schema or table is renamed; skipping this column removal.
    schemachanger_test.go:732: RENAME SCHEMA TO testsc_377
    schemachanger_test.go:783: ADD COLUMN col_376 TO testdb_316.testsc_349.t_345
    schemachanger_test.go:807: DROP COLUMN col_376 FROM testdb_316.testsc_377.t_345
    schemachanger_test.go:783: ADD COLUMN col_378 TO testdb_316.testsc_377.t_345
    schemachanger_test.go:783: ADD COLUMN col_381 TO testdb_316.testsc_377.t_345
    schemachanger_test.go:783: ADD COLUMN col_382 TO testdb_316.testsc_377.t_345
    schemachanger_test.go:807: DROP COLUMN col_381 FROM testdb_316.testsc_377.t_345
    schemachanger_test.go:807: DROP COLUMN col_378 FROM testdb_316.testsc_377.t_345
    schemachanger_test.go:762: DROP TABLE t_384
    schemachanger_test.go:762: DROP TABLE t_385
    schemachanger_test.go:783: ADD COLUMN col_386 TO testdb_316.testsc_377.t_345
    schemachanger_test.go:838: Parent database or schema or table is renamed or column is dropped or index already exists; skipping this index creation.
    schemachanger_test.go:783: ADD COLUMN col_387 TO testdb_316.testsc_377.t_345
    schemachanger_test.go:830: CREATE INDEX idx_388 ON testdb_316.testsc_377.t_345(col_386)
    schemachanger_test.go:856: DROP INDEX idx_388 FROM testdb_316.testsc_377.t_345
    schemachanger_test.go:734: DROP SCHEMA TO testsc_377
    schemachanger_test.go:811: Parent database or schema or table is renamed; skipping this column removal.
    schemachanger_test.go:783: ADD COLUMN col_390 TO testdb_316.testsc_377.t_345
    schemachanger_test.go:783: ADD COLUMN col_392 TO testdb_316.testsc_377.t_345
    schemachanger_test.go:830: CREATE INDEX idx_391 ON testdb_316.testsc_377.t_345(col_390)
    schemachanger_test.go:807: DROP COLUMN col_392 FROM testdb_316.testsc_377.t_345
    schemachanger_test.go:863: Parent database or schema or table is renamed or index is dropped; skipping this index removal.
    schemachanger_test.go:762: DROP TABLE t_393
    schemachanger_test.go:787: Parent database or schema or table is renamed or column already exists; skipping this column addition.
    schemachanger_test.go:760: RENAME TABLE TO t_394
    schemachanger_test.go:679: rename-db-worker encounters error driver: bad connection; signal to main routine and finish working
    schemachanger_test.go:675: rename-tbl-worker is signaled to finish work
    schemachanger_test.go:675: add-column-worker is signaled to finish work
    schemachanger_test.go:675: drop-index-worker is signaled to finish work
    schemachanger_test.go:675: drop-column-worker is signaled to finish work
    schemachanger_test.go:675: create-index-worker is signaled to finish work
    schemachanger_test.go:675: rename-schema-worker is signaled to finish work
    schemachanger_test.go:869: 
        	Error Trace:	pkg/sql/schemachanger/schemachanger_test_test/pkg/sql/schemachanger/schemachanger_test.go:869
        	Error:      	Received unexpected error:
        	            	driver: bad connection
        	Test:       	TestConcurrentSchemaChanges
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestConcurrentSchemaChanges1429434194
--- FAIL: TestConcurrentSchemaChanges (108.33s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=8</code>
 - <code>shard=24</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #129271 pkg/sql/schemachanger/schemachanger_test: TestConcurrentSchemaChanges failed [C-test-failure O-robot P-1 T-sql-foundations branch-release-24.2]
- #129220 pkg/sql/schemachanger/schemachanger_test: TestConcurrentSchemaChanges failed [C-test-failure O-robot P-1 T-sql-foundations branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConcurrentSchemaChanges.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/sql/schemachanger/schemachanger_test.go
+++ b/pkg/sql/schemachanger/schemachanger_test.go
@@ -620,10 +620,10 @@ func TestConcurrentSchemaChanges(t *testing.T) {
 		// Decrease the adopt loop interval so that retries happen quickly.
 		JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
 	}
-	s, sqlDB, _ := serverutils.StartServer(t, params)
+	s, setupConn, _ := serverutils.StartServer(t, params)
 	defer s.Stopper().Stop(ctx)
 	dbName, scName, tblName := "testdb", "testsc", "t"
-	useLegacyOrDeclarative := func() error {
+	useLegacyOrDeclarative := func(sqlDB *gosql.DB) error {
 		decl := rand.Intn(2) == 0
 		if !decl {
 			_, err := sqlDB.Exec("SET use_declarative_schema_changer='off';")
@@ -633,41 +633,41 @@ func TestConcurrentSchemaChanges(t *testing.T) {
 		return err
 	}
 
-	createSchema := func() error {
+	createSchema := func(conn *gosql.DB) error {
 		return testutils.SucceedsSoonError(func() error {
-			_, err := sqlDB.Exec(fmt.Sprintf("CREATE DATABASE IF NOT EXISTS %v;", dbName))
+			_, err := conn.Exec(fmt.Sprintf("CREATE DATABASE IF NOT EXISTS %v;", dbName))
 			if err != nil {
 				return err
 			}
-			_, err = sqlDB.Exec(fmt.Sprintf("CREATE SCHEMA IF NOT EXISTS %v.%v;", dbName, scName))
+			_, err = conn.Exec(fmt.Sprintf("CREATE SCHEMA IF NOT EXISTS %v.%v;", dbName, scName))
 			if err != nil {
 				return err
 			}
-			_, err = sqlDB.Exec(fmt.Sprintf("CREATE TABLE IF NOT EXISTS %v.%v.%v(col INT PRIMARY KEY);", dbName, scName, tblName))
+			_, err = conn.Exec(fmt.Sprintf("CREATE TABLE IF NOT EXISTS %v.%v.%v(col INT PRIMARY KEY);", dbName, scName, tblName))
 			if err != nil {
 				return err
 			}
-			_, err = sqlDB.Exec(fmt.Sprintf("DELETE FROM %v.%v.%v;", dbName, scName, tblName))
+			_, err = conn.Exec(fmt.Sprintf("DELETE FROM %v.%v.%v;", dbName, scName, tblName))
 			if err != nil {
 				return err
 			}
-			_, err = sqlDB.Exec(fmt.Sprintf("INSERT INTO %v.%v.%v SELECT generate_series(1,100);", dbName, scName, tblName))
+			_, err = conn.Exec(fmt.Sprintf("INSERT INTO %v.%v.%v SELECT generate_series(1,100);", dbName, scName, tblName))
 			if err != nil {
 				return err
 			}
 			return nil
 		})
 	}
-	require.NoError(t, createSchema())
+	require.NoError(t, createSchema(setupConn))
 
 	// repeatWorkWithInterval repeats `work` indefinitely every `workInterval` until
 	// `ctx` is cancelled.
 	repeatWorkWithInterval := func(
-		workerName string, workInterval time.Duration, work func(sqlDB *gosql.DB) error,
+		workerName string, workInterval time.Duration, work func(workConn *gosql.DB) error,
 	) func(context.Context) error {
 		return func(workerCtx context.Context) error {
-			sqlDB := s.SQLConn(t)
-			sqlDB.SetMaxOpenConns(1)
+			workConn := s.SQLConn(t)
+			workConn.SetMaxOpenConns(1)
 			for {
 				jitteredInterval := workInterval * time.Duration(0.8+0.4*rand.Float32())
 				select {
@@ -675,7 +675,7 @@ func TestConcurrentSchemaChanges(t *testing.T) {
 					t.Logf("%v is signaled to finish work", workerName)
 					return nil
 				case <-time.After(jitteredInterval):
-					if err := work(sqlDB); err != nil {
+					if err := work(workConn); err != nil {
 						t.Logf("%v encounters error %v; signal to main routine and finish working", workerName, err.Error())
 						return err
 					}
@@ -686,23 +686,23 @@ func TestConcurrentSchemaChanges(t *testing.T) {
 
 	var nextObjectID atomic.Int64
 	// A goroutine that repeatedly renames database `testdb` randomly.
-	g.GoCtx(repeatWorkWithInterval("rename-db-worker", renameDBInterval, func(sqlDB *gosql.DB) error {
-		if err := useLegacyOrDeclarative(); err != nil {
+	g.GoCtx(repeatWorkWithInterval("rename-db-worker", renameDBInterval, func(workerConn *gosql.DB) error {
+		if err := useLegacyOrDeclarative(workerConn); err != nil {
 			return err
 		}
 		drop := rand.Intn(2) == 0
 		if drop {
-			if _, err := sqlDB.Exec(fmt.Sprintf("DROP DATABASE %v CASCADE", dbName)); err != nil {
+			if _, err := workerConn.Exec(fmt.Sprintf("DROP DATABASE %v CASCADE", dbName)); err != nil {
 				return err
 			}
 			t.Logf("DROP DATABASE %v", dbName)
-			return createSchema()
+			return createSchema(workerConn)
 		}
 		newDBName := fmt.Sprintf("testdb_%v", nextObjectID.Add(1))
 		if newDBName == dbName {
 			return nil
 		}
-		if _, err := sqlDB.Exec(fmt.Sprintf("ALTER DATABASE %v RENAME TO %v", dbName, newDBName)); err != nil {
+		if _, err := workerConn.Exec(fmt.Sprintf("ALTER DATABASE %v RENAME TO %v", dbName, newDBName)); err != nil {
 			return err
 		}
 		dbName = newDBName
@@ -711,8 +711,8 @@ func TestConcurrentSchemaChanges(t *testing.T) {
 	}))
 
 	// A goroutine that renames schema `testdb.testsc` randomly.
-	g.GoCtx(repeatWorkWithInterval("rename-schema-worker", renameSCInterval, func(sqlDB *gosql.DB) error {
-		if err := useLegacyOrDeclarative(); err != nil {
+	g.GoCtx(repeatWorkWithInterval("rename-schema-worker", renameSCInterval, func(workerConn *gosql.DB) error {
+		if err := useLegacyOrDeclarative(workerConn); err != nil {
 			return err
 		}
 		drop := rand.Intn(2) == 0
@@ -722,9 +722,9 @@ func TestConcurrentSchemaChanges(t *testing.T) {
 		}
 		var err error
 		if !drop {
-			_, err = sqlDB.Exec(fmt.Sprintf("ALTER SCHEMA %v.%v RENAME TO %v", dbName, scName, newSCName))
+			_, err = workerConn.Exec(fmt.Sprintf("ALTER SCHEMA %v.%v RENAME TO %v", dbName, scName, newSCName))
 		} else {
-			_, err = sqlDB.Exec(fmt.Sprintf("DROP SCHEMA %v.%v CASCADE", dbName, scName))
+			_, err = workerConn.Exec(fmt.Sprintf("DROP SCHEMA %v.%v CASCADE", dbName, scName))
 		}
 		if err == nil {
 			if !drop {
@@ -732,7 +732,7 @@ func TestConcurrentSchemaChanges(t *testing.T) {
 				t.Logf("RENAME SCHEMA TO %v", newSCName)
 			} else {
 				t.Logf("DROP SCHEMA TO %v", scName)
-				return createSchema()
+				return createSchema(workerConn)
 			}
 		} else if isPQErrWithCode(err, pgcode.UndefinedDatabase, pgcode.UndefinedSchema) {
 			err = nil // mute those errors as they're expected
@@ -742,17 +742,17 @@ func TestConcurrentSchemaChanges(t *testing.T) {
 	}))
 
 	// A goroutine that renames table `testdb.testsc.t` randomly.
-	g.GoCtx(repeatWorkWithInterval("rename-tbl-worker", renameTblInterval, func(sqlDB *gosql.DB) error {
-		if err := useLegacyOrDeclarative(); err != nil {
+	g.GoCtx(repeatWorkWithInterval("rename-tbl-worker", renameTblInterval, func(workerConn *gosql.DB) error {
+		if err := useLegacyOrDeclarative(workerConn); err != nil {
 			return err
 		}
 		newTblName := fmt.Sprintf("t_%v", nextObjectID.Add(1))
 		drop := rand.Intn(2) == 0
 		var err error
 		if !drop {
-			_, err = sqlDB.Exec(fmt.Sprintf(`ALTER TABLE %v.%v.%v RENAME TO %v`, dbName, scName, tblName, newTblName))
+			_, err = workerConn.Exec(fmt.Sprintf(`ALTER TABLE %v.%v.%v RENAME TO %v`, dbName, scName, tblName, newTblName))
 		} else {
-			_, err = sqlDB.Exec(fmt.Sprintf(`DROP TABLE %v.%v.%v`, dbName, scName, tblName))
+			_, err = workerConn.Exec(fmt.Sprintf(`DROP TABLE %v.%v.%v`, dbName, scName, tblName))
 		}
 		if err == nil {
 			if !drop {
@@ -760,7 +760,7 @@ func TestConcurrentSchemaChanges(t *testing.T) {
 				t.Logf("RENAME TABLE TO %v", newTblName)
 			} else {
 				t.Logf("DROP TABLE %v", newTblName)
-				return createSchema()
+				return createSchema(workerConn)
 			}
 		} else if isPQErrWithCode(err, pgcode.UndefinedDatabase, pgcode.UndefinedSchema, pgcode.InvalidSchemaName, pgcode.UndefinedObject, pgcode.UndefinedTable) {
 			err = nil
@@ -770,14 +770,14 @@ func TestConcurrentSchemaChanges(t *testing.T) {
 	}))
 
 	// A goroutine that adds columns to `testdb.testsc.t` randomly.
-	g.GoCtx(repeatWorkWithInterval("add-column-worker", addColInterval, func(sqlDB *gosql.DB) error {
-		if err := useLegacyOrDeclarative(); err != nil {
+	g.GoCtx(repeatWorkWithInterval("add-column-worker", addColInterval, func(workerConn *gosql.DB) error {
+		if err := useLegacyOrDeclarative(workerConn); err != nil {
 			return err
 		}
 		dbName, scName, tblName := dbName, scName, tblName
 		newColName := fmt.Sprintf("col_%v", nextObjectID.Add(1))
 
-		_, err := sqlDB.Exec(fmt.Sprintf("ALTER TABLE %v.%v.%v ADD COLUMN %v INT DEFAULT %v",
+		_, err := workerConn.Exec(fmt.Sprintf("ALTER TABLE %v.%v.%v ADD COLUMN %v INT DEFAULT %v",
 			dbName, scName, tblName, newColName, rand.Intn(100)))
 		if err == nil {
 			t.Logf("ADD COLUMN %v TO %v.%v.%v", newColName, dbName, scName, tblName)
@@ -790,18 +790,18 @@ func TestConcurrentSchemaChanges(t *testing.T) {
 	}))
 
 	// A goroutine that drops columns from `testdb.testsc.t` randomly.
-	g.GoCtx(repeatWorkWithInterval("drop-column-worker", dropColInterval, func(sqlDB *gosql.DB) error {
-		if err := useLegacyOrDeclarative(); err != nil {
+	g.GoCtx(repeatWorkWithInterval("drop-column-worker", dropColInterval, func(workerConn *gosql.DB) error {
+		if err := useLegacyOrDeclarative(workerConn); err != nil {
 			return err
 		}
 		// Randomly pick a non-PK column to drop.
 		dbName, scName, tblName := dbName, scName, tblName
-		colName, err := getANonPrimaryKeyColumn(sqlDB, dbName, scName, tblName)
+		colName, err := getANonPrimaryKeyColumn(workerConn, dbName, scName, tblName)
 		if err != nil || colName == "" {
 			return err
 		}
 
-		_, err = sqlDB.Exec(fmt.Sprintf("ALTER TABLE %v.%v.%v DROP COLUMN %v;",
+		_, err = workerConn.Exec(fmt.Sprintf("ALTER TABLE %v.%v.%v DROP COLUMN %v;",
 			dbName, scName, tblName, colName))
 		if err == nil {
 			t.Logf("DROP COLUMN %v FROM %v.%v.%v", colName, dbName, scName, tblName)
@@ -814,17 +814,17 @@ func TestConcurrentSchemaChanges(t *testing.T) {
 	}))
 
 	// A goroutine that creates secondary index on a randomly selected column.
-	g.GoCtx(repeatWorkWithInterval("create-index-worker", createIdxInterval, func(sqlDB *gosql.DB) error {
+	g.GoCtx(repeatWorkWithInterval("create-index-worker", createIdxInterval, func(workerConn *gosql.DB) error {
 		newIndexName := fmt.Sprintf("idx_%v", nextObjectID.Add(1))
 
 		// Randomly pick a non-PK column to create an index on.
 		dbName, scName, tblName := dbName, scName, tblName
-		colName, err := getANonPrimaryKeyColumn(sqlDB, dbName, scName, tblName)
+		colName, err := getANonPrimaryKeyColumn(workerConn, dbName, scName, tblName)
 		if err != nil || colName == "" {
 			return err
 		}
 
-		_, err = sqlDB.Exec(fmt.Sprintf("CREATE INDEX %v ON %v.%v.%v (%v);",
+		_, err = workerConn.Exec(fmt.Sprintf("CREATE INDEX %v ON %v.%v.%v (%v);",
 			newIndexName, dbName, scName, tblName, colName))
 		if err == nil {
 			t.Logf("CREATE INDEX %v ON %v.%v.%v(%v)", newIndexName, dbName, scName, tblName, colName)
@@ -841,17 +841,17 @@ func TestConcurrentSchemaChanges(t *testing.T) {
 	}))
 
 	// A goroutine that drops a secondary index randomly.
-	g.GoCtx(repeatWorkWithInterval("drop-index-worker", dropIdxInterval, func(sqlDB *gosql.DB) error {
-		if err := useLegacyOrDeclarative(); err != nil {
+	g.GoCtx(repeatWorkWithInterval("drop-index-worker", dropIdxInterval, func(workerConn *gosql.DB) error {
+		if err := useLegacyOrDeclarative(workerConn); err != nil {
 			return err
 		}
 		// Randomly pick a public, secondary index to drop.
 		dbName, scName, tblName := dbName, scName, tblName
-		indexName, err := getASecondaryIndex(sqlDB, dbName, scName, tblName)
+		indexName, err := getASecondaryIndex(workerConn, dbName, scName, tblName)
 		if err != nil || indexName == "" {
 			return err
 		}
-		_, err = sqlDB.Exec(fmt.Sprintf("DROP INDEX %v.%v.%v@%v;", dbName, scName, tblName, indexName))
+		_, err = workerConn.Exec(fmt.Sprintf("DROP INDEX %v.%v.%v@%v;", dbName, scName, tblName, indexName))
 		if err == nil {
 			t.Logf("DROP INDEX %v FROM %v.%v.%v", indexName, dbName, scName, tblName)
 		} else if isPQErrWithCode(err, pgcode.UndefinedDatabase, pgcode.UndefinedSchema,
@@ -870,8 +870,8 @@ func TestConcurrentSchemaChanges(t *testing.T) {
 }
 
 // getANonPrimaryKeyColumn returns a non-primary-key column from table `dbName.scName.tblName`.
-func getANonPrimaryKeyColumn(sqlDB *gosql.DB, dbName, scName, tblName string) (string, error) {
-	colNameRow, err := sqlDB.Query(fmt.Sprintf(`
+func getANonPrimaryKeyColumn(workerConn *gosql.DB, dbName, scName, tblName string) (string, error) {
+	colNameRow, err := workerConn.Query(fmt.Sprintf(`
 SELECT column_name 
 FROM [show columns from %s.%s.%s] 
 WHERE column_name != 'col'
@@ -894,8 +894,8 @@ ORDER BY random();  -- shuffle column output
 }
 
 // getASecondaryIndex returns a secondary index from table `dbName.scName.tblName`.
-func getASecondaryIndex(sqlDB *gosql.DB, dbName, scName, tblName string) (string, error) {
-	colNameRow, err := sqlDB.Query(fmt.Sprintf(`
+func getASecondaryIndex(workerConn *gosql.DB, dbName, scName, tblName string) (string, error) {
+	colNameRow, err := workerConn.Query(fmt.Sprintf(`
 SELECT index_name 
 FROM [show indexes from %s.%s.%s]
 WHERE index_name NOT LIKE '%%_pkey'

